output/src/lib.rs

```diff
@@ -0,0 +1,152 @@
+use crate::columns::{OutputCols, NUM_OUTPUT_COLS, OUTPUT_COL_MAP};
+use core::iter;
+use core::mem::transmute;
+use valida_bus::MachineWithGeneralBus;
+use valida_cpu::MachineWithCpuChip;
+use valida_machine::{
+    instructions, BusArgument, Chip, Instruction, Interaction, Operands, Word, CPU_MEMORY_CHANNELS,
+    MEMORY_CELL_BYTES,
+};
+
+use p3_air::VirtualPairCol;
+use p3_field::PrimeField;
+use p3_matrix::dense::RowMajorMatrix;
+use p3_maybe_rayon::*;
+
+const WRITE_OPCODE: u32 = 102;
+
+pub mod columns;
+pub mod stark;
+
+pub struct OutputChip {
+    pub values: Vec<(u32, u8)>, // (clk, byte)
+}
+
+impl<F, M> Chip<M> for OutputChip
+where
+    F: PrimeField,
+    M: MachineWithGeneralBus<F = F>,
+{
+    fn generate_trace(&self, _machine: &M) -> RowMajorMatrix<M::F> {
+        let table_len = self.values.len() as u32;
+        let mut rows = self
+            .values
+            .as_slice()
+            .par_windows(2)
+            .map(|window| {
+                let clk_1 = window[0].0;
+                let clk_2 = window[1].0;
+                let val_1 = window[0].1;
+                let clk_diff = clk_2 - clk_1;
+                let num_rows = (clk_diff / table_len) as usize + 1;
```

Comment:
I think it'd be better if can react to increases in `table_len` as we add padding. Wouldn't matter in most cases, but it'd ensure the number of padding rows is O(sqrt(cycles)), vs possibly O(cycles) if there was a tiny output with a large clock diff.

Reply:
Agreed. Unless I'm missing something obvious, it's difficult to do this at the moment without refactoring the code and adding an initial serial pass. I'll merge for now, but feel free to refactor.